File: Python/136.Single Number.py
#一个数字数组，只有一个数字出现一次，其他数字出现了两次，找出该数字
#思想：每个数和自身的异或是0，所以从头异或一遍，结果为单个数
#或者利用set,把所有数全都缩减为1次，做差即可
from functools import reduce
import operator

class Solution(object):
    def singleNumber(self, nums):
        return reduce(operator.xor,nums)

    def singleNumber1(self, nums):
        result = nums[0]
        for i in range(1,len(nums)):
            result = result ^ nums[i]
        return result

    def singleNumber2(self,nums):
        return 2*sum(set(nums))-sum(nums)

s = Solution()
print(s.singleNumber([1,2,1,3,2]))

File: C++/Hash Table/136.Single Number.cpp
/*
数组中除了一个数其他每个数出现两次，找到该数。
*/

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
	int singleNumber(vector<int>& nums) {
        int result = 0;
        for(int n:nums)
        {
        	result ^= n;
        }
        return result;
    }
};

int main()
{
	Solution s;
	vector<int> nums{1,2,3,3,2,10,11,15,10,11,1};
	cout<<s.singleNumber(nums)<<endl;
}

